File: CBLClient/Apps/CBLTestServer-Dotnet/TestServer/ReplicatorTcpConnection.cs
// 
//  ReplicatorTcpConnection.cs
// 
//  Copyright (c) 2018 Couchbase, Inc All rights reserved.
// 
//  Licensed under the Couchbase License Agreement (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//  https://info.couchbase.com/rs/302-GJY-034/images/2017-10-30_License_Agreement.pdf
// 
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.

using System;
using System.Linq;
using System.Text;
using System.Text.RegularExpressions;
using System.Threading.Tasks;

using System.Threading;
using System.Net.Sockets;
using System.IO;
using System.Security.Cryptography;
using Couchbase.Lite.P2P;

namespace Couchbase.Lite.Testing
{
    /// <summary>
    /// This class is P2P Replicator Tcp communication implementing IMessageEndpointConnection
    /// </summary>
    public sealed class ReplicatorTcpConnection : IMessageEndpointConnection
    {
        #region Constants

        internal const int Port = 5000;
        private const int ReceiveBufferSize = 8192;

        #endregion

        #region Variables

        private readonly CancellationTokenSource _cancelSource = new CancellationTokenSource();
        private bool _connected;
        //sprivate LogHelper _log = new LogHelper(nameof(ReplicatorTcpConnection));
        private IReplicatorConnection _replicatorConnection;
        TcpClient _client;
        Stream _networkStream; //NetworkStream, SslStream

        #endregion

        #region Constructors

        /// <summary>
        /// Constructs a new replicator Active Peer connection instance via Tcp Connection
        /// </summary>
        /// <param name="client">tcp client that is accepted by the listener</param>
        public ReplicatorTcpConnection(TcpClient client)
        {
            // _client is *already* connected
            _client = client;
            _networkStream = _client.GetStream();
        }

        #endregion

        #region Private Methods

        private async void ReceiveLoop()
        {
            var buffer = new byte[ReceiveBufferSize];
            try
            {
                while (!_cancelSource.IsCancellationRequested)
                {
                    //read from _networkStream 				
                    int length;
                    while ((length = await _networkStream.ReadAsync(buffer, 0, buffer.Length).ConfigureAwait(false)) != 0)
                    {
                        _replicatorConnection.Receive(
                                Message.FromBytes(buffer.Take(length).ToArray()));
                    }
                }
            }
            catch (Exception)
            {
                //placerholder for customized logging or error handling
            }
        }

        #endregion

        #region IMessageEndpointConnection -- Passive Peer

        /// <summary>
        /// Closes the connection to the remote endpoint
        /// </summary>
        /// <param name="error">The error that caused the replication to close, if any</param>
        /// <returns>A task that can be awaited to determine the end of the close</returns>
        public async Task Close(Exception error)
        {
            if (!_connected)
            {
                return;
            }
            _cancelSource.Cancel();
            _connected = false;
            _client.Close();
            _client.Dispose();
            _client = null;
        }

        /// <summary>
        /// Opens a connection to the remote endpoint, linking it with the given
        /// <see cref="IReplicatorConnection"/>
        /// </summary>
        /// <param name="connection">The counterpart in the send/receive relationship which
        /// is used to signal the replication process of certain key events</param>
        /// <returns>A task that can be awaited to determine the end of the open</returns>
        public async Task Open(IReplicatorConnection connection)
        {
            if (_connected)
            {
                return;
            }
            _networkStream.PerformServerWebSocketHandshake();
            _connected = true;
            _replicatorConnection = connection;
            await Task.Factory.StartNew(ReceiveLoop, TaskCreationOptions.LongRunning).ConfigureAwait(false);
            Console.Write("Finished open");
        }

        /// <summary>
        /// Sends a message to the remote endpoint
        /// </summary>
        /// <param name="message">The message to send</param>
        /// <returns>A task that can be awaited to determine the end of the send</returns>
        public async Task Send(Message message)
        {
            var bytes = message.ToByteArray();
            await _networkStream.WriteAsync(bytes, 0, bytes.Length).ConfigureAwait(false);
        }

        #endregion
    }

    /*internal static class StreamExt
    {
        internal static bool PerformServerWebSocketHandshake(this Stream stream)
        {
            var buffer = new Byte[8192];
            stream.Read(buffer, 0, buffer.Length);

            //translate bytes of request to string
            string data = Encoding.UTF8.GetString(buffer);
            if (new System.Text.RegularExpressions.Regex("^GET").IsMatch(data))
            {
                StringBuilder sb = new StringBuilder();
                const string eol = "\r\n"; // HTTP/1.1 defines the sequence CR LF as the end-of-line marker
                try
                {
                    var key = new Regex("Sec-WebSocket-Key: (.*)").Match(data).Groups[1].Value.Trim();
                    var protocol = new Regex("Sec-WebSocket-Protocol: (.*)").Match(data).Groups[1].Value.Trim();
                    var version = new Regex("Sec-WebSocket-Version: (.*)").Match(data).Groups[1].Value.Trim();
                    Byte[] response = Encoding.UTF8.GetBytes(sb.Append("HTTP/1.1 101 Switching Protocols").Append(eol)
                    .Append("Connection: Upgrade").Append(eol)
                    .Append("Upgrade: websocket").Append(eol)
                    .Append("Sec-WebSocket-Version: ").Append(version).Append(eol)
                    .Append("Sec-WebSocket-Protocol: ").Append(protocol).Append(eol)
                    .Append("Sec-WebSocket-Accept: ").Append(AcceptKey(key)).Append(eol)
                    .Append(eol).ToString());
                    stream.Write(response, 0, response.Length);
                    return true;
                }
                catch (Exception)
                {
                    // failed to get regex matched value or any other exception
                    // placerholder for customized logging or error handling
                    return false;
                }
            }
            return false;
        }

         private static string AcceptKey(string key)
        {
            string longKey = key + "258EAFA5-E914-47DA-95CA-C5AB0DC85B11";
            SHA1 sha1 = SHA1CryptoServiceProvider.Create();
            byte[] hashBytes = sha1.ComputeHash(System.Text.Encoding.ASCII.GetBytes(longKey));
            return Convert.ToBase64String(hashBytes);
        } 
    } */
}

File: CBLClient/Apps/JavaRequestHandlers/src/main/java/com/couchbase/mobiletestkit/javacommon/RequestHandler/ReplicatorTcpConnection.java
package com.couchbase.mobiletestkit.javacommon.RequestHandler;

//
// ReplicatorTcpConnection.java
//
// Copyright (c) 2018 Couchbase, Inc.  All rights reserved.
//
// Licensed under the Couchbase License Agreement (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// https://info.couchbase.com/rs/302-GJY-034/images/2017-10-30_License_Agreement.pdf
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.Socket;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.Arrays;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import com.couchbase.mobiletestkit.javacommon.RequestHandlerDispatcher;
import com.couchbase.mobiletestkit.javacommon.util.Log;
import com.couchbase.lite.Message;
import com.couchbase.lite.MessageEndpointConnection;
import com.couchbase.lite.MessagingCloseCompletion;
import com.couchbase.lite.MessagingCompletion;
import com.couchbase.lite.MessagingError;
import com.couchbase.lite.ReplicatorConnection;


// Base ReplicatorTcpConnection that implements MessageEndpointConnection
public abstract class ReplicatorTcpConnection implements MessageEndpointConnection {
    protected static final int RECEIVE_BUFFER_SIZE = 8192;
    private static final String TAG = "CONN";

    protected Socket socket;
    protected InputStream inputStream;
    protected OutputStream outputStream;
    protected ReplicatorConnection replicatorConnection;

    private Thread receiveThread;
    private boolean connected;

    // Called by subclasses to set the socket when ready:
    protected void setSocket(Socket socket) throws IOException {
        if (this.socket != null) {
            throw new IllegalStateException("Socket has already been set.");
        }
        this.socket = socket;
        this.inputStream = socket.getInputStream();
        this.outputStream = socket.getOutputStream();
    }

    // Called by open(ReplicatorConnection, MessagingCompletion) for subclasses
    // to establish a connection to the remote peer:
    abstract boolean openConnection() throws Exception;

    // MessageEndpointConnection:

    @Override
    public void open(ReplicatorConnection connection, MessagingCompletion completion) {
        if (connected) {
            return;
        }

        MessagingError error = null;
        try {
            if (!openConnection()) {
                error = new MessagingError(new RuntimeException("Failed to open connection"), false);
            }
        }
        catch (Exception e) {
            error = new MessagingError(e, false);
        }

        completion.complete(error == null, error);

        if (error != null) {
            closeSocket();
            return;
        }

        connected = true;
        replicatorConnection = connection;
        receiveThread = new Thread(new Runnable() {
            @Override
            public void run() {
                receiveLoop();
            }
        });
        receiveThread.start();
    }

    @Override
    public void close(Exception error, final MessagingCloseCompletion completion) {
        if (!connected) {
            return;
        }
        connected = false;
        closeSocket();
        completion.complete();
    }

    @Override
    public void send(Message message, MessagingCompletion completion) {
        byte[] bytes = message.toData();
        try {
            outputStream.write(bytes);
            completion.complete(true, null);
        }
        catch (IOException e) {
            completion.complete(false, new MessagingError(e, false));
        }
    }

    // Private Methods:

    private void closeSocket() {
        try {
            if (socket != null) {
                socket.close();
                socket = null;
            }
        }
        catch (IOException e) {
            Log.e(TAG, "DB closing socket", e);
        }
    }

    private void receiveLoop() {
        Exception error = null;
        byte[] buffer = new byte[RECEIVE_BUFFER_SIZE];
        try {
            int length;
            while ((length = inputStream.read(buffer)) != 0) {
                replicatorConnection.receive(Message.fromData(Arrays.copyOfRange(buffer, 0, length)));
            }
        }
        catch (Exception e) {
            if (!(e instanceof InterruptedException)) {
                error = e;
            }
        }
        replicatorConnection.close(error != null ? new MessagingError(error, false) : null);
    }

    // WebSocket utilities used by subclasses:

    protected String getHTTPHeader(String data, String key) {
        Matcher m = Pattern.compile(key + ": (.*)").matcher(data);
        if (m.find()) { return m.group(1); }
        else { return null; }
    }

    protected String getWebSocketAcceptKey(String key) {
        String longKey = key.trim() + "258EAFA5-E914-47DA-95CA-C5AB0DC85B11";
        MessageDigest md;
        try {
            md = MessageDigest.getInstance("SHA-1");
        }
        catch (NoSuchAlgorithmException e) {
            return null;
        }

        byte[] hashBytes = md.digest(longKey.getBytes(StandardCharsets.US_ASCII));
        return RequestHandlerDispatcher.context.encodeBase64(hashBytes);
    }
}